nixos/modules/programs/dconf.nix

```diff
@@ -1,55 +1,174 @@
 { config, lib, pkgs, ... }:
 
-with lib;
-
 let
   cfg = config.programs.dconf;
-  cfgDir = pkgs.symlinkJoin {
-    name = "dconf-system-config";
-    paths = map (x: "${x}/etc/dconf") cfg.packages;
-    postBuild = ''
-      mkdir -p $out/profile
-      mkdir -p $out/db
-    '' + (
-      concatStringsSep "\n" (
-        mapAttrsToList (
-          name: path: ''
-            ln -s ${path} $out/profile/${name}
-          ''
-        ) cfg.profiles
-      )
-    ) + ''
-      ${pkgs.dconf}/bin/dconf update $out/db
-    '';
-  };
+
+  # Compile keyfiles to Dconf DB
+  compileDconfDb = dir: pkgs.runCommand "dconf-db"
+    {
+      nativeBuildInputs = [ pkgs.dconf ];
+    } "dconf compile $out ${dir}";
+
+  # Recursively lock values in settings
+  recuriseLock = with lib; settings: locks: (mapAttrsRecursive
+    (path: value: optionalAttrs value
+      (mapAttrsRecursiveCond (value: !lib.gvariant.isGVariant value)
+        (path: value: true)
+        (foldl'
+          (root: name: if hasAttr "${name}" root then root."${name}" else
+          throw "${concatStringsSep "." (tail path)} doesn't exist in settings.")
+          { v = settings; }
+          path)))
+    { v = locks; }).v;
+
+  # Generate Dconf DB from attrset and keyfiles
+  mkDconfDb = val: compileDconfDb (pkgs.symlinkJoin {
+    name = "nixos-generated-dconf-keyfiles";
+    paths = [
+      (pkgs.writeTextDir "nixos-generated-dconf-keyfiles" (lib.generators.toDconfINI val.settings))
+      (pkgs.writeTextDir "locks/nixos-generated-dconf-locks" (lib.generators.toDconfLocks
+        (lib.recursiveUpdate (recuriseLock val.settings val.lockSubpaths) val.locks)))
+      val.keyfiles
+    ];
+  });
+
+  # Generate Dconf profile
+  mkDconfProfile = name: value: pkgs.writeTextDir "etc/dconf/profile/${name}" (
+    if lib.isDerivation value || lib.isPath value then
+      lib.readFile value
```

Comment:
We do not really allow import-from-derivation in Nixpkgs since it freezes Nix evaluation until the read file is built. Use `runCommand` with `mkdir` + `cp` instead of `writeTextDir` for this branch.

Reply:
Done. Why should we use cp instead of ln?

Reply:
I do not think it makes much of a difference. Mainly, when the link target is a derivation, it will end up in the output closure. This might be annoying when the package is large and nothing other than the dconf stuff is used from it but I do not think this will happen in real use cases.